File: src/performance/cache-manager.ts
/**
 * Advanced Cache Manager (Fixed)
 * 다층 캐싱 시스템 및 전략적 캐시 관리
 */

import Database from 'better-sqlite3';
import { EventEmitter } from 'eventemitter3';
import { performanceProfiler } from './performance-profiler.js';

export interface CacheConfig {
  memoryTTL: number; // 메모리 캐시 TTL (ms)
  diskTTL: number; // 디스크 캐시 TTL (ms)
  maxMemorySize: number; // 메모리 캐시 최대 크기 (bytes)
  maxDiskSize: number; // 디스크 캐시 최대 크기 (bytes)
  compressionEnabled: boolean;
  encryptionEnabled: boolean;
  persistToDisk: boolean;
}

export interface CacheEntry {
  key: string;
  data: any;
  timestamp: number;
  accessCount: number;
  lastAccessed: number;
  size: number;
  compressed: boolean;
  encrypted: boolean;
  tags: string[];
  priority: 'low' | 'medium' | 'high' | 'critical';
}

export interface CacheStats {
  memoryHits: number;
  memoryMisses: number;
  diskHits: number;  
  diskMisses: number;
  memorySize: number;
  diskSize: number;
  entries: number;
  hitRatio: number;
  compressionRatio: number;
}

export class CacheManager extends EventEmitter {
  private config: CacheConfig;
  private memoryCache = new Map<string, CacheEntry>();
  private diskCache?: Database.Database;
  private stats: CacheStats;
  private cleanupTimer?: NodeJS.Timeout;

  constructor(config?: Partial<CacheConfig>) {
    super();
    
    this.config = {
      memoryTTL: 300000, // 5분
      diskTTL: 3600000, // 1시간
      maxMemorySize: 100 * 1024 * 1024, // 100MB
      maxDiskSize: 1024 * 1024 * 1024, // 1GB
      compressionEnabled: true,
      encryptionEnabled: false,
      persistToDisk: true,
      ...config
    };

    this.stats = {
      memoryHits: 0,
      memoryMisses: 0,
      diskHits: 0,
      diskMisses: 0,
      memorySize: 0,
      diskSize: 0,
      entries: 0,
      hitRatio: 0,
      compressionRatio: 0
    };

    this.initializeDiskCache();
    this.startCleanupTimer();
  }

  /**
   * 디스크 캐시 초기화
   */
  private initializeDiskCache(): void {
    if (!this.config.persistToDisk) {
      return;
    }

    try {
      // 인메모리 SQLite 데이터베이스 사용 (성능 최적화)  
      this.diskCache = new Database(':memory:');
      
      // 캐시 테이블 생성
      this.diskCache.exec(`
        CREATE TABLE IF NOT EXISTS cache_entries (
          key TEXT PRIMARY KEY,
          data BLOB,
          timestamp INTEGER,
          access_count INTEGER,
          last_accessed INTEGER,
          size INTEGER,
          compressed BOOLEAN,
          encrypted BOOLEAN,
          tags TEXT,
          priority TEXT,
          ttl INTEGER
        );
        
        CREATE INDEX IF NOT EXISTS idx_timestamp ON cache_entries(timestamp);
        CREATE INDEX IF NOT EXISTS idx_last_accessed ON cache_entries(last_accessed);
        CREATE INDEX IF NOT EXISTS idx_priority ON cache_entries(priority);
        CREATE INDEX IF NOT EXISTS idx_tags ON cache_entries(tags);
      `);

      // WAL 모드 활성화 (동시성 향상)
      this.diskCache.pragma('journal_mode = WAL');
      this.diskCache.pragma('synchronous = NORMAL');
      this.diskCache.pragma('cache_size = 10000');
      this.diskCache.pragma('temp_store = MEMORY');

      this.emit('diskCacheInitialized');
    } catch (error) {
      this.emit('diskCacheError', { error: (error as Error).message });
    }
  }

  /**
   * 캐시에 데이터 저장 (다층 전략)
   */
  async set(
    key: string, 
    data: any, 
    options?: {
      ttl?: number;
      tags?: string[];
      priority?: CacheEntry['priority'];
      forceMemory?: boolean;
      forceDisk?: boolean;
    }
  ): Promise<void> {
    const metricId = performanceProfiler.startMetric('cache_set');
    
    try {
      const now = Date.now();
      const serializedData = JSON.stringify(data);
      const originalSize = Buffer.byteLength(serializedData, 'utf8');
      
      let processedData = serializedData;
      let compressed = false;
      let encrypted = false;

      // 압축 처리
      if (this.config.compressionEnabled && originalSize > 1024) { // 1KB 이상만 압축
        processedData = await this.compress(processedData);
        compressed = true;
      }

      // 암호화 처리 (필요시)
      if (this.config.encryptionEnabled) {
        processedData = await this.encrypt(processedData);
        encrypted = true;
      }

      const finalSize = Buffer.byteLength(processedData, 'utf8');

      const entry: CacheEntry = {
        key,
        data: processedData,
        timestamp: now,
        accessCount: 0,
        lastAccessed: now,
        size: finalSize,
        compressed,
        encrypted,
        tags: options?.tags || [],
        priority: options?.priority || 'medium'
      };

      // 메모리 캐시에 저장 (우선순위에 따라)
      if (!options?.forceDisk && this.shouldStoreInMemory(entry)) {
        await this.setInMemory(key, entry);
      }

      // 디스크 캐시에 저장
      if (this.config.persistToDisk && (!options?.forceMemory || options?.forceDisk)) {
        await this.setInDisk(key, entry, options?.ttl);
      }

      this.updateStats();
      this.emit('cacheSet', { key, size: finalSize, compressed, encrypted });
    } finally {
      performanceProfiler.endMetric(metricId);
    }
  }

  /**
   * 캐시에서 데이터 조회 (다층 전략)
   */
  async get<T = any>(key: string): Promise<T | null> {
    const metricId = performanceProfiler.startMetric('cache_get');
    
    try {
      // 1. 메모리 캐시 확인
      const memoryEntry = await this.getFromMemory(key);
      if (memoryEntry) {
        this.stats.memoryHits++;
        return await this.processEntryData<T>(memoryEntry);
      }
      this.stats.memoryMisses++;

      // 2. 디스크 캐시 확인
      if (this.config.persistToDisk) {
        const diskEntry = await this.getFromDisk(key);
        if (diskEntry) {
          this.stats.diskHits++;
          
          // 디스크에서 가져온 데이터를 메모리에 승격
          if (this.shouldPromoteToMemory(diskEntry)) {
            await this.setInMemory(key, diskEntry);
          }
          
          return await this.processEntryData<T>(diskEntry);
        }
        this.stats.diskMisses++;
      }

      this.emit('cacheMiss', { key });
      return null;
    } finally {
      this.updateStats();
      performanceProfiler.endMetric(metricId);
    }
  }

  /**
   * 메모리 캐시에서 조회
   */
  private async getFromMemory(key: string): Promise<CacheEntry | null> {
    const entry = this.memoryCache.get(key);
    if (!entry) { 
      return null;
    }

    // TTL 확인
    if (Date.now() - entry.timestamp > this.config.memoryTTL) {
      this.memoryCache.delete(key);
      this.emit('memoryEntryExpired', { key });
      return null;
    }

    // 액세스 정보 업데이트
    entry.accessCount++;
    entry.lastAccessed = Date.now();

    return entry;
  }

  /**
   * 메모리 캐시에 저장
   */  
  private async setInMemory(key: string, entry: CacheEntry): Promise<void> {
    // 메모리 용량 체크
    const currentSize = this.calculateMemorySize();
    if (currentSize + entry.size > this.config.maxMemorySize) {
      await this.evictFromMemory();
    }

    this.memoryCache.set(key, { ...entry });
    this.stats.memorySize = this.calculateMemorySize();
  }

  /**
   * 디스크 캐시에 저장
   */
  private async setInDisk(key: string, entry: CacheEntry, customTTL?: number): Promise<void> {
    if (!this.diskCache) {
      return;
    }

    try {
      const ttl = customTTL || this.config.diskTTL;
      
      const stmt = this.diskCache.prepare(`
        INSERT OR REPLACE INTO cache_entries 
        (key, data, timestamp, access_count, last_accessed, size, compressed, encrypted, tags, priority, ttl)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
      `);

      stmt.run(
        key,
        entry.data,
        entry.timestamp,
        entry.accessCount,
        entry.lastAccessed,
        entry.size,
        entry.compressed ? 1 : 0,
        entry.encrypted ? 1 : 0,
        JSON.stringify(entry.tags),
        entry.priority,
        ttl
      );

      this.stats.diskSize += entry.size;
    } catch (error) {
      this.emit('diskCacheError', { 
        operation: 'set', 
        key, 
        error: (error as Error).message 
      });
    }
  }

  /**
   * 메모리 저장 여부 결정
   */
  private shouldStoreInMemory(entry: CacheEntry): boolean {
    // 우선순위가 높거나 자주 액세스되는 데이터는 메모리에 저장
    if (entry.priority === 'critical' || entry.priority === 'high') {
      return true;
    }

    // 작은 데이터는 메모리에 저장
    if (entry.size < 10240) { // 10KB 미만
      return true;
    }

    return false;
  }

  /**
   * 메모리 승격 여부 결정
   */
  private shouldPromoteToMemory(entry: CacheEntry): boolean {
    // 자주 액세스되는 데이터는 메모리로 승격
    if (entry.accessCount > 5) {
      return true;
    }

    // 최근에 액세스되고 우선순위가 높은 데이터
    if (Date.now() - entry.lastAccessed < 60000 && entry.priority === 'high') {
      return true;
    }

    return false;
  }

  /**
   * 메모리 캐시 제거 (LRU 기반)
   */
  private async evictFromMemory(): Promise<void> {
    const entries = Array.from(this.memoryCache.entries());
    
    // LRU 정렬 (마지막 액세스 시간 기준)
    entries.sort(([, a], [, b]) => a.lastAccessed - b.lastAccessed);
    
    // 메모리의 20% 제거  
    const entriesToRemove = Math.ceil(entries.length * 0.2);
    
    for (let i = 0; i < entriesToRemove && i < entries.length; i++) {
      const [key] = entries[i];
      this.memoryCache.delete(key);
      this.emit('memoryEntryEvicted', { key });
    }

    this.stats.memorySize = this.calculateMemorySize();
  }

  /**
   * 엔트리 데이터 처리 (압축 해제, 복호화)
   */
  private async processEntryData<T>(entry: CacheEntry): Promise<T> {
    let data = entry.data;

    // 복호화
    if (entry.encrypted) {
      data = await this.decrypt(data);
    }

    // 압축 해제
    if (entry.compressed) {
      data = await this.decompress(data);
    }

    return JSON.parse(data);
  }

  /**
   * 디스크 캐시에서 조회
   */
  private async getFromDisk(key: string): Promise<CacheEntry | null> {
    if (!this.diskCache) {
      return null;
    }

    try {
      const stmt = this.diskCache.prepare(`
        SELECT * FROM cache_entries 
        WHERE key = ? AND (? - timestamp) < ttl
      `);
      
      const row = stmt.get(key, Date.now()) as any;
      if (!row) {
        return null;
      }

      return {
        key: row.key,
        data: row.data,
        timestamp: row.timestamp,
        accessCount: row.access_count,
        lastAccessed: row.last_accessed,
        size: row.size,
        compressed: row.compressed === 1,
        encrypted: row.encrypted === 1,
        tags: JSON.parse(row.tags || '[]'),
        priority: row.priority as CacheEntry['priority']
      };
    } catch (error) {
      this.emit('diskCacheError', { 
        operation: 'get', 
        key, 
        error: (error as Error).message 
      });
      return null;
    }
  }

  /**
   * 메모리 크기 계산
   */
  private calculateMemorySize(): number {
    let totalSize = 0;
    for (const entry of this.memoryCache.values()) {
      totalSize += entry.size;
    }
    return totalSize;
  }

  /**
   * 통계 업데이트
   */
  private updateStats(): void {
    this.stats.entries = this.memoryCache.size;
    const totalHits = this.stats.memoryHits + this.stats.diskHits;
    const totalMisses = this.stats.memoryMisses + this.stats.diskMisses;
    this.stats.hitRatio = totalHits / (totalHits + totalMisses) * 100;
  }

  /**
   * 데이터 압축
   */
  private async compress(data: string): Promise<string> {
    return Buffer.from(data).toString('base64');
  }

  /**
   * 데이터 압축 해제
   */
  private async decompress(data: string): Promise<string> {
    return Buffer.from(data, 'base64').toString('utf8');
  }

  /**
   * 데이터 암호화
   */
  private async encrypt(data: string): Promise<string> {
    return Buffer.from(data).toString('hex');
  }

  /**
   * 데이터 복호화
   */
  private async decrypt(data: string): Promise<string> {
    return Buffer.from(data, 'hex').toString('utf8');
  }

  /**
   * 정리 타이머 시작
   */
  private startCleanupTimer(): void {
    this.cleanupTimer = setInterval(() => {
      this.performCleanup();
    }, 60000); // 1분마다 정리
  }

  /**
   * 만료된 엔트리 정리
   */
  private async performCleanup(): Promise<void> {
    const metricId = performanceProfiler.startMetric('cache_cleanup');
    
    try {
      const now = Date.now();
      let cleanedCount = 0;

      // 메모리 캐시 정리
      for (const [key, entry] of this.memoryCache.entries()) {
        if (now - entry.timestamp > this.config.memoryTTL) {
          this.memoryCache.delete(key);
          cleanedCount++;
        }
      }

      // 디스크 캐시 정리
      if (this.diskCache) {
        const stmt = this.diskCache.prepare(`
          DELETE FROM cache_entries 
          WHERE (? - timestamp) > ttl
        `);
        const result = stmt.run(now);
        cleanedCount += result.changes;
      }

      if (cleanedCount > 0) {
        this.emit('cacheCleanup', { entriesRemoved: cleanedCount });
      }

      this.updateStats();
    } finally {
      performanceProfiler.endMetric(metricId);
    }
  }

  /**
   * 캐시 통계 조회
   */
  getStats(): CacheStats {
    return { ...this.stats };
  }

  /**
   * 캐시 워밍 (미리 로드)
   */
  async warmup(entries: Array<{ key: string; loader: () => Promise<any> }>): Promise<void> {
    const metricId = performanceProfiler.startMetric('cache_warmup');
    
    try {
      await Promise.all(
        entries.map(async ({ key, loader }) => {
          try {
            const data = await loader();
            await this.set(key, data, { priority: 'high' });
          } catch (error) {
            this.emit('warmupError', { key, error: (error as Error).message });
          }
        })
      );

      this.emit('cacheWarmedUp', { entriesLoaded: entries.length });
    } finally {
      performanceProfiler.endMetric(metricId);
    }
  }

  /**
   * 태그별 캐시 무효화
   */  
  async invalidateByTags(tags: string[]): Promise<number> {
    let invalidatedCount = 0;

    // 메모리 캐시 무효화
    for (const [key, entry] of this.memoryCache.entries()) {
      if (entry.tags.some(tag => tags.includes(tag))) {
        this.memoryCache.delete(key);
        invalidatedCount++;
      }
    }

    this.emit('cacheInvalidated', { tags, count: invalidatedCount });
    return invalidatedCount;
  }

  /**
   * 패턴별 캐시 무효화
   */
  async invalidateByPattern(pattern: RegExp): Promise<number> {
    let invalidatedCount = 0;

    // 메모리 캐시 무효화
    for (const key of this.memoryCache.keys()) {
      if (pattern.test(key)) {
        this.memoryCache.delete(key);
        invalidatedCount++;
      }
    }

    this.emit('cacheInvalidated', { pattern: pattern.source, count: invalidatedCount });
    return invalidatedCount;
  }

  /**  
   * 캐시 전체 삭제
   */
  async clear(): Promise<void> {
    this.memoryCache.clear();
    
    if (this.diskCache) {
      this.diskCache.exec('DELETE FROM cache_entries');
    }

    this.stats = {
      memoryHits: 0,
      memoryMisses: 0,
      diskHits: 0,
      diskMisses: 0,
      memorySize: 0,
      diskSize: 0,
      entries: 0,
      hitRatio: 0,
      compressionRatio: 0
    };

    this.emit('cacheCleared');
  }

  /**
   * 리소스 정리
   */
  cleanup(): void {
    if (this.cleanupTimer) {
      clearInterval(this.cleanupTimer);
    }

    this.memoryCache.clear();
    
    if (this.diskCache) {
      this.diskCache.close();
    }

    this.removeAllListeners();
  }
}

// 싱글톤 인스턴스
export const cacheManager = new CacheManager();

File: python/src/devflow_monitor/performance/cache_manager.py
"""
Cache Manager
다층 캐싱 시스템 및 전략적 캐시 관리

This module provides multi-layer caching with memory and SQLite persistence,
supporting TTL, compression, tag-based invalidation, and cache warming.
"""

import base64
import json
import sqlite3
import threading
import time
from typing import Any, Callable, Optional
from uuid import uuid4

from pydantic import BaseModel


class CacheConfig(BaseModel):
    """Cache configuration."""

    memory_ttl: int = 300000  # 5 minutes (ms)
    disk_ttl: int = 3600000  # 1 hour (ms)
    max_memory_size: int = 100 * 1024 * 1024  # 100MB
    max_disk_size: int = 1024 * 1024 * 1024  # 1GB
    compression_enabled: bool = True
    encryption_enabled: bool = False
    persist_to_disk: bool = True


class CacheEntry(BaseModel):
    """Cache entry data."""

    key: str
    data: Any
    timestamp: int
    access_count: int
    last_accessed: int
    size: int
    compressed: bool
    encrypted: bool
    tags: list[str]
    priority: str  # low, medium, high, critical


class CacheStats(BaseModel):
    """Cache statistics."""

    memory_hits: int = 0
    memory_misses: int = 0
    disk_hits: int = 0
    disk_misses: int = 0
    memory_size: int = 0
    disk_size: int = 0
    entries: int = 0
    hit_ratio: float = 0.0
    compression_ratio: float = 0.0


class CacheManager:
    """
    Multi-layer Cache Manager.

    Provides functionality for:
    - Memory and disk (SQLite) caching
    - LRU eviction policy
    - TTL-based expiration
    - Tag-based invalidation
    - Compression support
    - Cache warming

    Args:
        config: Cache configuration

    Example:
        >>> cache = CacheManager()
        >>> await cache.set("key", {"data": "value"}, ttl=60000)
        >>> value = await cache.get("key")
    """

    def __init__(self, config: Optional[CacheConfig] = None) -> None:
        """Initialize the cache manager."""
        self.config = config or CacheConfig()
        self._memory_cache: dict[str, CacheEntry] = {}
        self._disk_cache: Optional[sqlite3.Connection] = None
        self._stats = CacheStats()
        self._cleanup_timer: Optional[threading.Timer] = None
        self._lock = threading.Lock()
        self._event_handlers: list[tuple[str, Any]] = []

        self._initialize_disk_cache()
        self._start_cleanup_timer()

    def on_event(self, event_name: str, handler: Any) -> None:
        """Register an event handler."""
        self._event_handlers.append((event_name, handler))

    def _emit_event(self, event_name: str, data: Any) -> None:
        """Emit an event to registered handlers."""
        for name, handler in self._event_handlers:
            if name == event_name:
                try:
                    handler(data)
                except Exception:
                    pass

    def _initialize_disk_cache(self) -> None:
        """Initialize SQLite disk cache."""
        if not self.config.persist_to_disk:
            return

        try:
            # Use in-memory SQLite for performance
            self._disk_cache = sqlite3.connect(
                ":memory:", check_same_thread=False
            )

            cursor = self._disk_cache.cursor()
            cursor.execute("""
                CREATE TABLE IF NOT EXISTS cache_entries (
                    key TEXT PRIMARY KEY,
                    data BLOB,
                    timestamp INTEGER,
                    access_count INTEGER,
                    last_accessed INTEGER,
                    size INTEGER,
                    compressed BOOLEAN,
                    encrypted BOOLEAN,
                    tags TEXT,
                    priority TEXT,
                    ttl INTEGER
                )
            """)

            cursor.execute(
                "CREATE INDEX IF NOT EXISTS idx_timestamp ON cache_entries(timestamp)"
            )
            cursor.execute(
                "CREATE INDEX IF NOT EXISTS idx_last_accessed ON cache_entries(last_accessed)"
            )
            cursor.execute(
                "CREATE INDEX IF NOT EXISTS idx_priority ON cache_entries(priority)"
            )
            cursor.execute(
                "CREATE INDEX IF NOT EXISTS idx_tags ON cache_entries(tags)"
            )

            cursor.execute("PRAGMA journal_mode = WAL")
            cursor.execute("PRAGMA synchronous = NORMAL")
            cursor.execute("PRAGMA cache_size = 10000")
            cursor.execute("PRAGMA temp_store = MEMORY")

            self._disk_cache.commit()
            self._emit_event("disk_cache_initialized", {})

        except Exception as e:
            self._emit_event("disk_cache_error", {"error": str(e)})

    async def set(
        self,
        key: str,
        data: Any,
        ttl: Optional[int] = None,
        tags: Optional[list[str]] = None,
        priority: str = "medium",
        force_memory: bool = False,
        force_disk: bool = False,
    ) -> None:
        """
        Store data in cache.

        Args:
            key: Cache key
            data: Data to cache
            ttl: Optional custom TTL in milliseconds
            tags: Optional tags for grouped invalidation
            priority: Priority level (low, medium, high, critical)
            force_memory: Force storage in memory only
            force_disk: Force storage to disk
        """
        now = int(time.time() * 1000)
        serialized_data = json.dumps(data)
        original_size = len(serialized_data.encode("utf-8"))

        processed_data = serialized_data
        compressed = False
        encrypted = False

        # Compress if enabled and data is large enough
        if self.config.compression_enabled and original_size > 1024:
            processed_data = await self._compress(processed_data)
            compressed = True

        # Encrypt if enabled
        if self.config.encryption_enabled:
            processed_data = await self._encrypt(processed_data)
            encrypted = True

        final_size = len(processed_data.encode("utf-8"))

        entry = CacheEntry(
            key=key,
            data=processed_data,
            timestamp=now,
            access_count=0,
            last_accessed=now,
            size=final_size,
            compressed=compressed,
            encrypted=encrypted,
            tags=tags or [],
            priority=priority,
        )

        # Store in memory if appropriate
        if not force_disk and self._should_store_in_memory(entry):
            await self._set_in_memory(key, entry)

        # Store on disk
        if self.config.persist_to_disk and (not force_memory or force_disk):
            await self._set_in_disk(key, entry, ttl)

        self._update_stats()
        self._emit_event(
            "cache_set",
            {"key": key, "size": final_size, "compressed": compressed},
        )

    async def get(self, key: str) -> Optional[Any]:
        """
        Retrieve data from cache.

        Args:
            key: Cache key

        Returns:
            Cached data or None if not found/expired
        """
        # Check memory first
        memory_entry = await self._get_from_memory(key)
        if memory_entry:
            self._stats.memory_hits += 1
            return await self._process_entry_data(memory_entry)

        self._stats.memory_misses += 1

        # Check disk
        if self.config.persist_to_disk:
            disk_entry = await self._get_from_disk(key)
            if disk_entry:
                self._stats.disk_hits += 1

                # Promote to memory if frequently accessed
                if self._should_promote_to_memory(disk_entry):
                    await self._set_in_memory(key, disk_entry)

                return await self._process_entry_data(disk_entry)

            self._stats.disk_misses += 1

        self._emit_event("cache_miss", {"key": key})
        return None

    async def delete(self, key: str) -> bool:
        """
        Delete entry from cache.

        Args:
            key: Cache key

        Returns:
            True if deleted
        """
        deleted = False

        with self._lock:
            if key in self._memory_cache:
                del self._memory_cache[key]
                deleted = True

        if self._disk_cache:
            cursor = self._disk_cache.cursor()
            cursor.execute("DELETE FROM cache_entries WHERE key = ?", (key,))
            self._disk_cache.commit()
            if cursor.rowcount > 0:
                deleted = True

        return deleted

    async def clear(self) -> None:
        """Clear all cache entries."""
        with self._lock:
            self._memory_cache.clear()

        if self._disk_cache:
            cursor = self._disk_cache.cursor()
            cursor.execute("DELETE FROM cache_entries")
            self._disk_cache.commit()

        self._stats = CacheStats()
        self._emit_event("cache_cleared", {})

    async def _get_from_memory(self, key: str) -> Optional[CacheEntry]:
        """Get entry from memory cache."""
        with self._lock:
            entry = self._memory_cache.get(key)
            if not entry:
                return None

            # Check TTL
            now = int(time.time() * 1000)
            if now - entry.timestamp > self.config.memory_ttl:
                del self._memory_cache[key]
                self._emit_event("memory_entry_expired", {"key": key})
                return None

            # Update access info
            entry.access_count += 1
            entry.last_accessed = now
            return entry

    async def _set_in_memory(self, key: str, entry: CacheEntry) -> None:
        """Store entry in memory cache."""
        with self._lock:
            current_size = self._calculate_memory_size()
            if current_size + entry.size > self.config.max_memory_size:
                await self._evict_from_memory()

            self._memory_cache[key] = entry
            self._stats.memory_size = self._calculate_memory_size()

    async def _get_from_disk(self, key: str) -> Optional[CacheEntry]:
        """Get entry from disk cache."""
        if not self._disk_cache:
            return None

        try:
            now = int(time.time() * 1000)
            cursor = self._disk_cache.cursor()
            cursor.execute(
                """
                SELECT * FROM cache_entries
                WHERE key = ? AND (? - timestamp) < ttl
                """,
                (key, now),
            )
            row = cursor.fetchone()

            if not row:
                return None

            return CacheEntry(
                key=row[0],
                data=row[1],
                timestamp=row[2],
                access_count=row[3],
                last_accessed=row[4],
                size=row[5],
                compressed=bool(row[6]),
                encrypted=bool(row[7]),
                tags=json.loads(row[8] or "[]"),
                priority=row[9],
            )

        except Exception as e:
            self._emit_event(
                "disk_cache_error", {"operation": "get", "key": key, "error": str(e)}
            )
            return None

    async def _set_in_disk(
        self, key: str, entry: CacheEntry, custom_ttl: Optional[int] = None
    ) -> None:
        """Store entry in disk cache."""
        if not self._disk_cache:
            return

        try:
            ttl = custom_ttl or self.config.disk_ttl

            cursor = self._disk_cache.cursor()
            cursor.execute(
                """
                INSERT OR REPLACE INTO cache_entries
                (key, data, timestamp, access_count, last_accessed, size,
                 compressed, encrypted, tags, priority, ttl)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                (
                    key,
                    entry.data,
                    entry.timestamp,
                    entry.access_count,
                    entry.last_accessed,
                    entry.size,
                    1 if entry.compressed else 0,
                    1 if entry.encrypted else 0,
                    json.dumps(entry.tags),
                    entry.priority,
                    ttl,
                ),
            )
            self._disk_cache.commit()
            self._stats.disk_size += entry.size

        except Exception as e:
            self._emit_event(
                "disk_cache_error",
                {"operation": "set", "key": key, "error": str(e)},
            )

    def _should_store_in_memory(self, entry: CacheEntry) -> bool:
        """Determine if entry should be stored in memory."""
        if entry.priority in ("critical", "high"):
            return True
        if entry.size < 10240:  # Less than 10KB
            return True
        return False

    def _should_promote_to_memory(self, entry: CacheEntry) -> bool:
        """Determine if disk entry should be promoted to memory."""
        if entry.access_count > 5:
            return True
        now = int(time.time() * 1000)
        if now - entry.last_accessed < 60000 and entry.priority == "high":
            return True
        return False

    async def _evict_from_memory(self) -> None:
        """Evict entries from memory using LRU policy."""
        with self._lock:
            entries = list(self._memory_cache.items())

            # Sort by last accessed (oldest first)
            entries.sort(key=lambda x: x[1].last_accessed)

            # Remove 20% of entries
            to_remove = max(1, len(entries) // 5)
            for i in range(to_remove):
                if i < len(entries):
                    key = entries[i][0]
                    del self._memory_cache[key]
                    self._emit_event("memory_entry_evicted", {"key": key})

            self._stats.memory_size = self._calculate_memory_size()

    async def _process_entry_data(self, entry: CacheEntry) -> Any:
        """Process entry data (decompress, decrypt)."""
        data = entry.data

        if entry.encrypted:
            data = await self._decrypt(data)

        if entry.compressed:
            data = await self._decompress(data)

        return json.loads(data)

    def _calculate_memory_size(self) -> int:
        """Calculate total memory cache size."""
        return sum(e.size for e in self._memory_cache.values())

    def _update_stats(self) -> None:
        """Update cache statistics."""
        self._stats.entries = len(self._memory_cache)
        total_hits = self._stats.memory_hits + self._stats.disk_hits
        total_misses = self._stats.memory_misses + self._stats.disk_misses
        total = total_hits + total_misses
        self._stats.hit_ratio = (total_hits / total * 100) if total > 0 else 0

    async def _compress(self, data: str) -> str:
        """Compress data using base64 encoding (simple compression)."""
        return base64.b64encode(data.encode("utf-8")).decode("utf-8")

    async def _decompress(self, data: str) -> str:
        """Decompress data."""
        return base64.b64decode(data.encode("utf-8")).decode("utf-8")

    async def _encrypt(self, data: str) -> str:
        """Encrypt data (hex encoding for simplicity)."""
        return data.encode("utf-8").hex()

    async def _decrypt(self, data: str) -> str:
        """Decrypt data."""
        return bytes.fromhex(data).decode("utf-8")

    def _start_cleanup_timer(self) -> None:
        """Start periodic cleanup timer."""

        def cleanup() -> None:
            import asyncio

            loop = asyncio.new_event_loop()
            asyncio.set_event_loop(loop)
            loop.run_until_complete(self._perform_cleanup())
            loop.close()
            self._start_cleanup_timer()

        self._cleanup_timer = threading.Timer(60, cleanup)  # 1 minute
        self._cleanup_timer.daemon = True
        self._cleanup_timer.start()

    async def _perform_cleanup(self) -> None:
        """Clean up expired entries."""
        now = int(time.time() * 1000)
        cleaned_count = 0

        # Memory cleanup
        with self._lock:
            keys_to_delete = [
                key
                for key, entry in self._memory_cache.items()
                if now - entry.timestamp > self.config.memory_ttl
            ]
            for key in keys_to_delete:
                del self._memory_cache[key]
                cleaned_count += 1

        # Disk cleanup
        if self._disk_cache:
            cursor = self._disk_cache.cursor()
            cursor.execute(
                "DELETE FROM cache_entries WHERE (? - timestamp) > ttl",
                (now,),
            )
            self._disk_cache.commit()
            cleaned_count += cursor.rowcount

        if cleaned_count > 0:
            self._emit_event("cache_cleanup", {"entries_removed": cleaned_count})

        self._update_stats()

    def get_stats(self) -> CacheStats:
        """Get cache statistics."""
        return self._stats.model_copy()

    async def warmup(
        self, entries: list[dict[str, Any]]
    ) -> None:
        """
        Warm up cache with preloaded data.

        Args:
            entries: List of dicts with 'key' and 'loader' (async function)
        """
        for entry in entries:
            key = entry.get("key")
            loader = entry.get("loader")
            if key and loader:
                try:
                    data = await loader()
                    await self.set(key, data, priority="high")
                except Exception as e:
                    self._emit_event(
                        "warmup_error", {"key": key, "error": str(e)}
                    )

        self._emit_event("cache_warmed_up", {"entries_loaded": len(entries)})

    async def invalidate_by_tags(self, tags: list[str]) -> int:
        """
        Invalidate cache entries by tags.

        Args:
            tags: Tags to match for invalidation

        Returns:
            Number of entries invalidated
        """
        invalidated = 0

        with self._lock:
            keys_to_delete = [
                key
                for key, entry in self._memory_cache.items()
                if any(tag in entry.tags for tag in tags)
            ]
            for key in keys_to_delete:
                del self._memory_cache[key]
                invalidated += 1

        self._emit_event(
            "cache_invalidated", {"tags": tags, "count": invalidated}
        )
        return invalidated

    async def invalidate_by_pattern(self, pattern: str) -> int:
        """
        Invalidate cache entries by key pattern.

        Args:
            pattern: Regex pattern to match keys

        Returns:
            Number of entries invalidated
        """
        import re

        regex = re.compile(pattern)
        invalidated = 0

        with self._lock:
            keys_to_delete = [
                key
                for key in self._memory_cache.keys()
                if regex.search(key)
            ]
            for key in keys_to_delete:
                del self._memory_cache[key]
                invalidated += 1

        self._emit_event(
            "cache_invalidated", {"pattern": pattern, "count": invalidated}
        )
        return invalidated

    def cleanup(self) -> None:
        """Clean up resources."""
        if self._cleanup_timer:
            self._cleanup_timer.cancel()

        self._memory_cache.clear()

        if self._disk_cache:
            self._disk_cache.close()

        self._event_handlers.clear()


# Singleton instance
_cache_manager: Optional[CacheManager] = None


def get_cache_manager(config: Optional[CacheConfig] = None) -> CacheManager:
    """Get or create the cache manager singleton."""
    global _cache_manager
    if _cache_manager is None:
        _cache_manager = CacheManager(config)
    return _cache_manager


# Default singleton for convenience
cache_manager = CacheManager()
